mediatek: Ruijie RG-X60 Pro: add OpenWrt UBI layout

airjinkela opened the pull request:
mediatek: add support for Ruijie RG-X60 Pro (UBI layout)

This commit adds support for Ruijie RG-X60 Pro (UBI layout).

OpenWrt U-Boot UBI flash instructions
-------------------------------------

Back up critical data
---------------------
  1. While the device is running OpenWrt with stock MTD partitions:
     ```BASH
     mtd dump Factory > /tmp/Factory.bin
     mtd dump product_info > /tmp/product_info.bin
     mtd dump BL2 > /tmp/BL2.bin
     mtd dump FIP > /tmp/FIP.bin
     mtd dump u-boot-env > /tmp/u-boot-env.bin
     ```
  3. Copy the backup files to your PC via SCP.

Create a new Factory image
--------------------------
Use this python script to create a new factory image
```python
FACTORY_IMAGE = "Factory.bin"
PRODUCT_INFO_IMAGE = "product_info.bin"
NEW_FACRORY_IMAGE = "Factory_new.bin"
def main():
  with (open(PRODUCT_INFO_IMAGE, "rb") as fp,
        open(FACTORY_IMAGE, "rb") as ff,
        open(NEW_FACRORY_IMAGE, "wb") as fn):
    groups_str = [part.decode('utf-8', errors="ignore") for part in fp.read()[4:].split(b'\x00') if part]
    mac_bytes = next((bytes.fromhex(p.split('=')[1].replace(':', '')) for p in groups_str if p.startswith('ethaddr=')), b"")
    if mac_bytes is None or len(mac_bytes) != 6:
      print("Error: cannot found ethaddr in product_info")
      return
    data = bytearray(ff.read())
    data[0x24:0x24 + 6] = mac_bytes
    fn.write(data)
main()

```

Using the installer image
-------------------------
1. Download openwrt-mediatek-filogic-ruijie_rg-x60-pro-initramfs-ubi-install.bin
   from https://github.com/airjinkela/owrt-misc/tree/master/ruijie-rg-x60pro-ubi

2. Force flash openwrt-mediatek-filogic-ruijie_rg-x60-pro-initramfs-ubi-install.bin
   via sysupgrade.

4. Upload files Via SCP:
   Factory_new.bin
   openwrt*ruijie_rg-x60-pro-ubi-preloader.bin
   openwrt*ruijie_rg-x60-pro-ubi-bl31-uboot.fip
   openwrt*ruijie_rg-x60-pro-ubi-initramfs-recovery.itb

5. Connect via SSH/UART.
6. Prepare the factory file:
   ```BASH
   dd if=/tmp/Factory_new.bin of=/tmp/factory bs=4096 count=1
   ```
7. Prepare UBI:
   ```BASH
   mtd write /tmp/*preloader.bin BL2
   ubidetach -p /dev/mtd1 ; ubiformat -y /dev/mtd1 ; ubiattach -p /dev/mtd1
   ubimkvol /dev/ubi0 -n 0 -t static -s $(du -b /tmp/*uboot.fip | awk '{print $1}') -N fip
   ubiupdatevol /dev/ubi0_0 /tmp/*uboot.fip
   ubimkvol /dev/ubi0 -n 1 -t static -s 4096 -N factory
   ubiupdatevol /dev/ubi0_1 /tmp/factory
   ubimkvol /dev/ubi0 -n 2 -s 126976 -N ubootenv
   ubimkvol /dev/ubi0 -n 3 -s 126976 -N ubootenv2
   ubimkvol /dev/ubi0 -n 4 -s $(du -b /tmp/*recovery.itb | awk '{print $1}') -N recovery
   ubiupdatevol /dev/ubi0_4 /tmp/*recovery.itb
   ```
8. Reboot the device to recovery mode:
   ```BASH
   reboot
   ```
9. Flash openwrt*ruijie_rg-x60-pro-ubi-squashfs-sysupgrade.itb
   via sysupgrade.

Return to stock MTD
-------------------
1. Force flash openwrt*ruijie_rg-x60-pro-initramfs-kernel.bin
   via sysupgrade
2. Copy files to /tmp on the device via SCP:
   BL2.bin
   Factory.bin
   FIP.bin
   product_info.bin
   openwrt*ruijie_rg-x60-pro-squashfs-sysupgrade.bin
3. Restore stock MTD partitions:
   ```BASH
   apk add kmod-mtd-rw
   insmod mtd-rw i_want_a_brick=1
   mtd write /tmp/BL2.bin BL2
   mtd write /tmp/FIP.bin FIP
   mtd write /tmp/u-boot-env.bin u-boot-env
   mtd write /tmp/Factory.bin Factory
   mtd write /tmp/product_info.bin product_info
   ```
4. Install the system:
   ```BASH
   sysupgrade /tmp/*sysupgrade.bin
   ```

BL2 and FIP recovery
--------------------
Use mtk_uartboot and UART connection if BL2 or FIP is destroyed
Link: https://github.com/981213/mtk_uartboot

Stock layout
```
-----------------------------------------
| dev:    size   erasesize  name        |
| mtd0: 00100000 00020000 "BL2"         |
| mtd1: 00080000 00020000 "u-boot-env"  |
| mtd2: 00200000 00020000 "Factory"     |
| mtd3: 00200000 00020000 "FIP"         |
| mtd4: 00080000 00020000 "product_info"|
| mtd5: 00080000 00020000 "kdump"       |
| mtd6: 03f00000 00020000 "ubi"         |
-----------------------------------------
```
OpenWrt U-Boot UBI layout
```
----------------------------------
| dev:    size   erasesize  name |
| mtd0: 00100000 00020000 "BL2"  |
| mtd1: 07f00000 00020000 "ubi"  |
----------------------------------
```

aiamadeus commented on package/boot/uboot-mediatek/patches/464-add-ruijie-rg-x60-pro.patch:
```diff
@@ -0,0 +1,371 @@
++
++	keys {
++		compatible = "gpio-keys";
++		factory {
```
`reset`, keep a blank line.

aiamadeus commented on package/boot/uboot-mediatek/patches/464-add-ruijie-rg-x60-pro.patch:
```diff
@@ -0,0 +1,371 @@
++		spi-max-frequency = <20000000>;
++		spi-tx-buswidth = <4>;
++		spi-rx-buswidth = <4>;
++		partitions {
```
Keep a blank line.

aiamadeus commented on package/boot/uboot-mediatek/patches/464-add-ruijie-rg-x60-pro.patch:
```diff
@@ -0,0 +1,371 @@
++		compatible = "spi-nand";
++		reg = <0>;
++		spi-max-frequency = <20000000>;
++		spi-tx-buswidth = <4>;
```
```
		spi-tx-bus-width = <4>;
		spi-rx-bus-width = <4>;
```

aiamadeus commented on package/boot/uboot-mediatek/patches/464-add-ruijie-rg-x60-pro.patch:
```diff
@@ -0,0 +1,371 @@
++&watchdog {
++	status = "disabled";
++};
++
```
An extra blank line at the end?

aiamadeus commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dtsi:
```diff
@@ -0,0 +1,333 @@
+		compatible = "mediatek,eth-mac";
+		reg = <0>;
+		phy-mode = "2500base-x";
+		fixed-link {
```
Keep a blank line.

aiamadeus commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1673,6 +1673,30 @@ define Device/ruijie_rg-x60-pro
 
+define Device/ruijie_rg-x60-pro-ubootmod
+  DEVICE_VENDOR := Ruijie
+  DEVICE_MODEL := RG-X60 Pro (OpenWrt U-Boot layout)
```
`DEVICE_VARIANT := (OpenWrt U-Boot layout)`

airjinkela replied:
OK, ALL FIXED

airjinkela committed:
mediatek: add support for Ruijie RG-X60 Pro (UBI layout)

This commit adds support for Ruijie RG-X60 Pro (UBI layout).

OpenWrt U-Boot UBI flash instructions
-------------------------------------

Back up critical data
---------------------
1. While the device is running OpenWrt with stock MTD partitions:
   mtd dump Factory > /tmp/Factory.bin
   mtd dump product_info > /tmp/product_info.bin
   mtd dump BL2 > /tmp/BL2.bin
   mtd dump FIP > /tmp/FIP.bin
   mtd dump u-boot-env > /tmp/u-boot-env.bin
2. Copy the backup files to your PC via SCP.

Create a new Factory image
--------------------------
Use this python script to create a new factory image
```python
FACTORY_IMAGE = "Factory.bin"
PRODUCT_INFO_IMAGE = "product_info.bin"
NEW_FACRORY_IMAGE = "Factory_new.bin"
def main():
  with (open(PRODUCT_INFO_IMAGE, "rb") as fp,
        open(FACTORY_IMAGE, "rb") as ff,
        open(NEW_FACRORY_IMAGE, "wb") as fn):
    groups_str = [part.decode('utf-8', errors="ignore") for part in fp.read()[4:].split(b'\x00') if part]
    mac_bytes = next((bytes.fromhex(p.split('=')[1].replace(':', '')) for p in groups_str if p.startswith('ethaddr=')), b"")
    if mac_bytes is None or len(mac_bytes) != 6:
      print("Error: cannot found ethaddr in product_info")
      return
    data = bytearray(ff.read())
    data[0x24:0x24 + 6] = mac_bytes
    fn.write(data)
main()

```

Using the installer image
-------------------------
1. Download openwrt-mediatek-filogic-ruijie_rg-x60-pro-initramfs-ubi-install.bin
   from https://github.com/airjinkela/owrt-misc/tree/master/ruijie-rg-x60pro-ubi
2. Force flash openwrt-mediatek-filogic-ruijie_rg-x60-pro-initramfs-ubi-install.bin
   via sysupgrade.
3. Upload files Via SCP:
   Factory_new.bin
   openwrt*ruijie_rg-x60-pro-ubi-preloader.bin
   openwrt*ruijie_rg-x60-pro-ubi-bl31-uboot.fip
   openwrt*ruijie_rg-x60-pro-ubi-initramfs-recovery.itb
4. Connect via SSH/UART.
5. Prepare the factory file:
   dd if=/tmp/Factory_new.bin of=/tmp/factory bs=4096 count=1
6. Prepare UBI:
   mtd write /tmp/*preloader.bin BL2
   ubidetach -p /dev/mtd1 ; ubiformat -y /dev/mtd1 ; ubiattach -p /dev/mtd1
   ubimkvol /dev/ubi0 -n 0 -t static -s $(du -b /tmp/*uboot.fip | awk '{print $1}') -N fip
   ubiupdatevol /dev/ubi0_0 /tmp/*uboot.fip
   ubimkvol /dev/ubi0 -n 1 -t static -s 4096 -N factory
   ubiupdatevol /dev/ubi0_1 /tmp/factory
   ubimkvol /dev/ubi0 -n 2 -s 126976 -N ubootenv
   ubimkvol /dev/ubi0 -n 3 -s 126976 -N ubootenv2
   ubimkvol /dev/ubi0 -n 4 -s $(du -b /tmp/*recovery.itb | awk '{print $1}') -N recovery
   ubiupdatevol /dev/ubi0_4 /tmp/*recovery.itb
7. Reboot the device to recovery mode:
   reboot
8. Flash openwrt*ruijie_rg-x60-pro-ubi-squashfs-sysupgrade.itb
   via sysupgrade.

Return to stock MTD
-------------------
1. Force flash openwrt*ruijie_rg-x60-pro-initramfs-kernel.bin
   via sysupgrade
2. Copy files to /tmp on the device via SCP:
   BL2.bin
   Factory.bin
   FIP.bin
   product_info.bin
   openwrt*ruijie_rg-x60-pro-squashfs-sysupgrade.bin
3. Restore stock MTD partitions:
   apk add kmod-mtd-rw
   insmod mtd-rw i_want_a_brick=1
   mtd write /tmp/BL2.bin BL2
   mtd write /tmp/FIP.bin FIP
   mtd write /tmp/u-boot-env.bin u-boot-env
   mtd write /tmp/Factory.bin Factory
   mtd write /tmp/product_info.bin product_info
4. Install the system:
   sysupgrade /tmp/*sysupgrade.bin

BL2 and FIP recovery
--------------------
Use mtk_uartboot and UART connection if BL2 or FIP is destroyed
Link: https://github.com/981213/mtk_uartboot

Stock layout
-----------------------------------------
| dev:    size   erasesize  name        |
| mtd0: 00100000 00020000 "BL2"         |
| mtd1: 00080000 00020000 "u-boot-env"  |
| mtd2: 00200000 00020000 "Factory"     |
| mtd3: 00200000 00020000 "FIP"         |
| mtd4: 00080000 00020000 "product_info"|
| mtd5: 00080000 00020000 "kdump"       |
| mtd6: 03f00000 00020000 "ubi"         |
-----------------------------------------

OpenWrt U-Boot UBI layout
----------------------------------
| dev:    size   erasesize  name |
| mtd0: 00100000 00020000 "BL2"  |
| mtd1: 07f00000 00020000 "ubi"  |
----------------------------------

Signed-off-by: air jinkela <air_jinkela@163.com>

airjinkela commented:
@aiamadeus review please thanks

aiamadeus commented on package/boot/uboot-mediatek/patches/464-add-ruijie-rg-x60-pro.patch:
```diff
@@ -0,0 +1,372 @@
+@@ -0,0 +1,180 @@
++/dts-v1/;
++#include <dt-bindings/input/linux-event-codes.h>
++#include "mt7986.dtsi"
```
```
+#include "mt7986.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/linux-event-codes.h>
```

airjinkela replied:
OK, fix

airjinkela commented:
@aiamadeus  Is the code good now

airjinkela commented:
@aiamadeus 
Is there anything else I need modify in this PR?

airjinkela commented:
> Just wait for the maintainer.

ok ,thank your help

abajk commented on package/boot/uboot-mediatek/patches/464-add-ruijie-rg-x60-pro.patch:
```diff
@@ -0,0 +1,372 @@
++	spi_nand@0 {
++		compatible = "spi-nand";
++		reg = <0>;
++		spi-max-frequency = <20000000>;
```
On most boards 52MHz works fine. Can it go higher?

airjinkela commented:
@hauke review please ,thank you

airjinkela commented:
Ping @hauke 
need review ,please

airjinkela commented:
@DragonBluep  Done

airjinkela commented:
@DragonBluep review please

airjinkela commented:
@hauke 
@DragonBluep
I may have been forgotten  (Sad)

airjinkela commented:
@hauke
@DragonBluep

DragonBluep commented on package/boot/uboot-mediatek/patches/465-add-ruijie-rg-x60-pro.patch:
```diff
@@ -0,0 +1,356 @@
++		compatible = "gpio-leds";
++
++		led-0 {
++			label = "white:status";
```
Please use `function` and `color` properties instead of `labels`.

DragonBluep commented on package/boot/uboot-mediatek/patches/465-add-ruijie-rg-x60-pro.patch:
```diff
@@ -0,0 +1,356 @@
++		};
++	};
++
++	uart1_pins: spi1-pins-func-3 {
```
Remove this unused node.

DragonBluep commented on package/boot/uboot-mediatek/patches/465-add-ruijie-rg-x60-pro.patch:
```diff
@@ -0,0 +1,356 @@
++CONFIG_BOOTDELAY=30
++CONFIG_AUTOBOOT_KEYED=y
++CONFIG_AUTOBOOT_MENU_SHOW=y
++CONFIG_DEFAULT_FDT_FILE="mediatek/mt7986a-ruijie_rg-x60-pro.dtb"
```
```suggestion
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro-ubootmod.dts:
```diff
@@ -0,0 +1,28 @@
+
+&chosen {
+	rootdisk = <&ubi_rootdisk>;
+	bootargs = " root=/dev/fit0 rootwait";
```
It's better to move this bootargs into the bootloader. Ref: https://git.openwrt.org/?p=openwrt%2Fopenwrt.git&a=search&h=HEAD&st=grep&s=root%3D%2Fdev%2Ffit0+rootwait

DragonBluep reviewed (commented):
@airjinkela Hi! Please check these comments. Sorry for the late.

airjinkela committed:
mediatek: increase SPI max frequency for Ruijie RG-X60 Pro

Update `spi-max-frequency` from 20 MHz to 52 MHz to better match
hardware capabilities and improve SPI performance.

Signed-off-by: air jinkela <air_jinkela@163.com>

DragonBluep commented:
LGTM for the patch diff changes. Please fix the commit message CI warnings:
```
[warn] Commit body line 4 is longer than 75 characters (is 99):
2. Get maximum available free space in OpenWrt ,The stock layout wasted approximately 44M of space.
[warn] Commit body line 92 is longer than 75 characters (is 104):
5. Power on the device and quickly press "up" "down" on the keyboard, then U-Boot will stay in the menu.
[warn] Commit body line 96 is longer than 75 characters (is 127):
7. Input the IP address, input the Stock “E-WEBOS” image file name to be flashed, start the TFTP server, and press "Enter".
```

airjinkela commented:
@DragonBluep All CI  have passed now

csharper2005 commented:
Would you like to make `ubi` layout for OpenWrt U-Boot as recommended here?https://github.com/openwrt/openwrt/pull/20666#discussion_r2988585767

csharper2005 commented:
> This is not user-friendly for ordinary users

What could be simpler for ordinary users?
```
1. Navigate http://192.168.1.1/ and download mtd backups
2. Upgrade OpenWrt with installer initramfs image (force upgrade, don't
   keep settings). Wait until OpenWrt reboots and until installer:
   - Prepare new factory partition
   - Format new ubi
   - Make ubi volumes
   - Write new fip and bl2
3. Navigate http://192.168.1.1/ and Upgrade with OpenWrt 'sysupgrade.bin'
   image (don't keep settings)
```

P.S. I won't insist.

airjinkela commented:
> > I will write a better commit msg later
> 
> Are you planning to drop non-ubi ubootmod layout from this PR later?

OK droped
